crates/adapters/src/controller/error.rs

```diff
@@ -211,6 +257,9 @@ impl Display for ControllerError {
             Self::DbspError { error } => {
                 write!(f, "DBSP error: '{error}'")
             }
+            Self::PipelineTerminating => {
+                f.write_str("Operation failed to complete because the pipeline is shutting down")
```

Comment:
If this error is getting bubbled to the pipeline manager (and beyond), it would be good to keep the error self contained and also state which pipeline this is referring to.

Reply:
Re this and previous comment: I am not convinced that self-contained errors are the right direction.  An error always occurs in the context of a specific request, which includes pipeline id and a bunch of other parameters, so it shouldn't be necessary to duplicate all this information inside the error itself (which would be required to make the error fully self contained).  On the other hand, there's some extra complexity associated with creating self-contained errors, as all the context information would have to be available when generating the error.  It's all doable, I just feel like we'd be going overboard in the wrong direction.  Thoughts?

Reply:
Yeah, it's tricky to balance. 

I'm thinking of this from the POV of looking at log messages. For example, we routinely do things like `debug!("XYZ operation failed: reason '{}'", error)` in code. So in a case like this, we'd just get the message "reason: 'Operation failed to complete because the pipeline is shutting down'".  Having the error message be self-contained enough here might help ("...because the pipeline foobar/<some-uuid> is shutting down...").

I think it'll be hard to make messages fully self contained as you say (request ID etc), but we could keep it best effort. I'd think an error like `PipelineTerminating` is likely generated at points in the code that know what pipeline/pipeline-id is being referenced?

Reply:
> I think it'll be hard to make messages fully self contained as you say (request ID etc), but we could keep it best effort. I'd think an error like PipelineTerminating is likely generated at points in the code that know what pipeline/pipeline-id is being referenced?

It gets generated inside the pipeline process itself, where all log messages are already annotated with pipeline id. But the pipeline id may not be readily available at the location where the error is geneated, so we'd have to store it in a global variable or pass as an argument to various functions, which would introduce unnecessary complexity.